awx/api/serializers.py

```diff
@@ -3375,20 +3406,38 @@ def validate(self, attrs):
         JT_job_tags = obj.job_tags
         JT_skip_tags = obj.skip_tags
         JT_inventory = obj.inventory
-        JT_credential = obj.credential
         JT_verbosity = obj.verbosity
-        extra_credentials = attrs.pop('extra_credentials', None)
+        credentials = attrs.pop('credentials', None)
         attrs = super(JobLaunchSerializer, self).validate(attrs)
         obj.extra_vars = JT_extra_vars
         obj.limit = JT_limit
         obj.job_type = JT_job_type
         obj.skip_tags = JT_skip_tags
         obj.job_tags = JT_job_tags
         obj.inventory = JT_inventory
-        obj.credential = JT_credential
         obj.verbosity = JT_verbosity
-        if extra_credentials is not None:
-            attrs['extra_credentials'] = extra_credentials
+        if credentials is not None:
+            attrs['credentials'] = credentials
+
+        # if the POST includes a list of credentials, verify that they don't
+        # require launch-time passwords
+        # if the POST *does not* include a list of credentials, fall back to
+        # checking the credentials on the JobTemplate
+        credentials = attrs['credentials'] if 'credentials' in data else obj.credentials.all()
+        passwords_needed = []
+        for cred in credentials:
+            if cred.passwords_needed:
+                passwords = self.context.get('passwords')
+                try:
+                    for p in cred.passwords_needed:
+                        passwords[p] = data[p]
+                except KeyError:
+                    passwords_needed.extend(cred.passwords_needed)
+        if len(passwords_needed):
+            raise serializers.ValidationError({
+                'passwords_needed_to_start': passwords_needed
+            })
+
```

Comment:
This is a very logically tortured loop. I asked myself "how does relaunch do it?"

Turns out, that's broken, see #685

While job template launch isn't broken, the variable handling is more than a little painful. On my prior question about how the credentials magically change from integers into objects, it was the previous credential validation loop in the serializers, operating on `data`, which is in the context, comes from the view, and is a copy of `request.data`, which has already been modified (forced mutable). That is made even more complicated by the direct editing of `passwords` inside of the context here. The method that purportedly validates the answers actually fills the critical role of filling out the values in the first place. I don't feel like any newcomer to the code could reasonably be expected to track down this variable flow.